tests/cxx/elaborated_type.cc

```diff
@@ -60,6 +60,21 @@ void global_qualified_elab(class ::GlobalClass* c);
 // they must also be forward-declared.
 Elaboration::Template<int, float>* namespace_qualified_template;
 
+// Elaborated types can be treated as forward-declarations, hence no fwd-decl is
+// needed if a declaration is introduced by elaborated type.
+// A declaration "owned" by elaborated type is an implicit declaration
+// introduced by elaborated type if no one was already present.
+void owning_elaborated(class OwningElaborated*);
+
+class OwningElaborated;
+void previously_introduced(OwningElaborated*);
+
+// But if an elaborated type appears after the first fwd-decl use of the type, a
+// true forward declaration is still needed.
+class ForNonOwningElaborated;
```

Comment:
I wonder if we should skip the whole owning/non-owning terminology... Isn't a regular forward-declaration, such as `class ForNonOwningElaborated;` really also an elaborated type? cppreference explains it that way: https://en.cppreference.com/w/cpp/language/elaborated_type_specifier. But I suppose Clang does not model the elaborated type specifier here, because the declaration introduces the type anyway.

From a language user's point of view, the rule seems to be more like:

> If an elaborated type specifier does not have a previous declaration of the same kind + name, it _is_ a declaration (in global scope)

Maybe something like `FirstElaborated` and `FirstForwardDeclared` would get the point across with less compiler internals?

Reply:
IWYU behavior probably should not depend on whether an elaborated type introduces an implicit declaration or not (e.g. because a declaration is already visible in some indirect header), despite it is not still the case. So I'm biased to remove all these subtle details from the test.

Reply:
I think the "owning" terminology is internal to Clang, so I'd at least tone that down.

But it seems useful with tests to help answer the question;
```
Why does IWYU remove my forward declaration of Bar in this scenario???

// fwd-decls
void foo(class Bar*);
class Bar;

void baz() {
    return foo(nullptr);
}
```
The "elaborated-type-introduces-a-decl" rule is subtle enough that someone will run into it and be surprised.